File: proxyshield/proxyshield-core/internal/middleware/honeypot.go
package middleware

import (
	"net/http"
	"sync"
	"time"

	"github.com/tejaspatil1936/proxyshield-core/internal/config"
	"github.com/tejaspatil1936/proxyshield-core/internal/event"
	"github.com/tejaspatil1936/proxyshield-core/internal/reqctx"
)

// Honeypot bans any IP that accesses a configured trap URL.
type Honeypot struct {
	config *config.Config
	banMap *sync.Map
}

// NewHoneypot creates a Honeypot middleware.
func NewHoneypot(cfg *config.Config, banMap *sync.Map) *Honeypot {
	return &Honeypot{config: cfg, banMap: banMap}
}

// Name returns the middleware identifier.
func (m *Honeypot) Name() string { return "honeypot" }

// Handle bans the client IP if it accesses any configured honeypot path.
func (m *Honeypot) Handle(w http.ResponseWriter, r *http.Request, ctx *reqctx.Context) bool {
	path := r.URL.Path

	for _, hp := range m.config.Honeypots {
		if path != hp.Path {
			continue
		}

		banDuration := time.Duration(hp.BanMinutes) * time.Minute
		m.banMap.Store(ctx.IP, BanEntry{
			BannedAt:    time.Now(),
			BanDuration: banDuration,
		})

		ctx.EventBus.Publish(event.Event{
			Name: event.IPBanned,
			Data: map[string]interface{}{
				"ip": ctx.IP, "path": path, "banMinutes": hp.BanMinutes,
			},
			Timestamp: time.Now(),
		})
		ctx.EventBus.Publish(event.Event{
			Name: event.RequestBlocked,
			Data: map[string]interface{}{
				"ip": ctx.IP, "path": path, "threatTag": "HONEYPOT_TRAP",
			},
			Timestamp: time.Now(),
		})

		writeJSON(w, http.StatusForbidden, map[string]string{
			"error":  "Forbidden",
			"reason": "HONEYPOT_TRAP",
		})
		return true
	}

	return false
}

File: proxyshield/src/middlewares/honeypot.js
/**
 * Honeypot middleware factory.
 * Defines fake trap URLs that no legitimate user would visit.
 * When a request hits a honeypot path, auto-ban the IP for a configured duration.
 *
 * @param {Map<string, { bannedAt: number, banDurationMs: number }>} runtimeBanMap - Shared runtime ban map
 * @returns {Function} Middleware function with signature (req, res, context) => Promise<string>
 */
export function createHoneypot(runtimeBanMap) {
  return async function honeypot(req, res, context) {
    // Extract just the pathname (without query string)
    let pathname;
    try {
      pathname = new URL(req.url, 'http://localhost').pathname;
    } catch {
      pathname = req.url;
    }

    // Check if the pathname matches any honeypot path
    const honeypots = context.config.honeypots || [];
    const match = honeypots.find(hp => hp.path === pathname);

    if (!match) return 'next';

    // Calculate ban duration
    const banDurationMs = match.ban_minutes * 60 * 1000;

    // Ban the IP
    runtimeBanMap.set(context.ip, {
      bannedAt: Date.now(),
      banDurationMs
    });

    // Emit events
    context.eventBus.emit('ip:banned', {
      ip: context.ip,
      path: pathname,
      banMinutes: match.ban_minutes,
      timestamp: Date.now()
    });

    context.eventBus.emit('request:blocked', {
      ip: context.ip,
      method: req.method,
      path: req.url,
      reason: 'Honeypot trap triggered',
      threatTag: 'HONEYPOT_TRAP',
      timestamp: Date.now()
    });

    res.writeHead(403, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'Forbidden', reason: 'HONEYPOT_TRAP' }));

    return 'stop';
  };
}
